sql/src/main/java/io/crate/operation/collect/collectors/ShardStateAwareRemoteCollector.java

```diff
@@ -0,0 +1,218 @@
+                                          int shardId,
+                                          RowConsumer consumer,
+                                          ClusterService clusterService,
+                                          IndicesService indicesService,
+                                          Function<IndexShard, CrateCollector> localCollectorProvider,
+                                          Function<String, RemoteCollector> remoteCollectorProvider,
+                                          ExecutorService executorService,
+                                          ThreadContext threadContext) {
+        this.indexName = indexName;
+        this.shardId = shardId;
+        this.consumer = consumer;
+        this.clusterService = clusterService;
+        this.localNode = clusterService.localNode().getId();
+        this.indicesService = indicesService;
+        this.localCollectorProvider = localCollectorProvider;
+        this.remoteCollectorProvider = remoteCollectorProvider;
+        this.executorService = executorService;
+        this.threadContext = threadContext;
+    }
+
+    @Override
+    public void doCollect() {
+        ClusterState clusterState = clusterService.state();
+        checkStateAndCollect(clusterState);
+    }
+
+    private void checkStateAndCollect(ClusterState clusterState) {
+        IndexShardRoutingTable shardRoutings;
+        try {
+            shardRoutings = clusterState.routingTable().shardRoutingTable(indexName, shardId);
+        } catch (ShardNotFoundException | IndexNotFoundException e) {
+            consumer.accept(null, e);
+            return;
+        }
+
+        // for update operations primaryShards must be used
+        // (for others that wouldn't be the case, but at this point it is not easily visible which is the case)
+        ShardRouting shardRouting = shardRoutings.primaryShard();
+        String remoteNode = shardRouting.currentNodeId();
+        if (localNode.equals(remoteNode) == false) {
+            synchronized (killLock) {
+                if (collectorKilled == false) {
+                    collector = remoteCollectorProvider.apply(remoteNode);
+                } else {
+                    consumer.accept(null, new InterruptedException());
+                }
+            }
+            // collect outside the synchronized block as the collectors handle kill signals that arrive while collecting
+            if (collector != null) {
+                collector.doCollect();
+            }
+        } else {
+            if (shardRouting.started()) {
+                // The shard is on the local node, so don't create a remote collector but a local one.
+                // We might've arrived here as a result of the cluster state observer detecting the shard relocated, in
+                // which case the ClusterState-Update thread is running this.
+                // The creation of a ShardCollectorProvider accesses the clusterState, which leads to an assertionError
```

Comment:
Just curious: Did you consider/look into making the clusterState a parameter for the `ShardCollectorProvider`, so that this thread dispatching wouldn't be necessary?